### PR TITLE
Added Parser

### PR DESCRIPTION
**Pull request by vkobinski**
Added a lot of Python AST Parsing to Bend Books.

### CODE REVIEW

File: crates/benda/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+crate-type = ["cdylib"]
+
+[dependencies]
+pyo3 = { version = "0.21.2", features = ["multiple-pymethods"] }
+bend-lang = "0.2.33"
+num-bigint = "0.4.5"
+num-traits = "0.2.19"
+rustpython-parser = "0.3.1"
+indexmap = "2.2.3"
+interner = "0.2.1"
```
Comment (steinerkelvin) [inline crates/benda/Cargo.toml]: Please add comments to non-obvious dependencies.

(We are always striving to have fewer dependencies, so it's good to have the non-obvious ones explained)

File: crates/benda/src/parser/builtins.rs [hunk 2]
```diff
@@ -0,0 +1,3 @@
+pub enum BuiltIn {
```
Comment (steinerkelvin) [inline crates/benda/src/parser/builtins.rs:1]: What is this supposed to be?

File: crates/benda/src/types/mod.rs [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        }
+    }
+}
+
+impl BendType for PyFloat {
+    fn to_bend(&self) -> imp::Expr {
+        let num: f32 = self.extract().unwrap();
+        imp::Expr::Num { val: Num::F24(num) }
+    }
+}
```
Comment (steinerkelvin) [inline crates/benda/src/types/mod.rs]: A cast from `f32` to `F24` is an operation that could fail, so I imagine `BendType::to_bend` should actually return a Result akin to std::convert::[TryInto](https://doc.rust-lang.org/std/convert/trait.TryInto.html#).

Actually, the whole type `BendType` could be `TryInto<Expr>` if it's only method will be `to_bend`.

File: crates/benda/src/types/mod.rs [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+pub enum BuiltinType {
+    U24,
+    F32,
+    I32,
+    Tree,
+    Leaf,
+    Node,
+}
+
+impl From<String> for BuiltinType {
```
Comment (steinerkelvin) [inline crates/benda/src/types/mod.rs:66]: Is this being used anywhere?

I don't think we should be relying on strings like this in the final version.
(if we can rely on in the identity of the types though the Python API and Python itself is not using strings)

File: crates/benda/src/types/u24.rs [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        u24::new(self.0 - other.0)
+    }
+}
+
+#[pymethods]
+impl u24 {
+    #[new]
+    fn new_py(value: u32) -> Self {
+        u24::new(value)
+    }
```
Comment (steinerkelvin) [inline crates/benda/src/types/u24.rs:82]: Add TODO comments for properly implementing the operations (masking, overflow etc) and tests.

We should have TODO comments in the points on the code we already know are lacking something.

File: crates/benda/src/parser/mod.rs [hunk 6]
```diff
@@ -0,0 +1,958 @@
...
+    }
+
+    pub fn get_var_name(&self) -> Option<Name> {
+        if let FromExpr::Expr(Expr::Var { nam }) = self {
+            Some(nam.clone())
+        } else {
+            None
+        }
+    }
+}
```
Comment (steinerkelvin) [inline crates/benda/src/parser/mod.rs:46]: Looks redundant 🤔

File: crates/benda/Cargo.toml (hunk 1 above)
Comment (vkobinski) [reply crates/benda/Cargo.toml]: IndexMap is used by Bend to store their ADTs and constructor definitions, so I needed to add it to be able to construct a new IndexMap when parsing a new ADT.

File: crates/benda/src/parser/builtins.rs (hunk 2 above)
Comment (vkobinski) [reply crates/benda/src/parser/builtins.rs:1]: I was trying to figure out a syntax for a switch with binding, but then I went through with another idea. Dead code until we pin out a better syntax for a switch with binding https://github.com/fglab-tech/benda/issues/6.

File: crates/benda/src/types/mod.rs (hunk 3 above)
Comment (vkobinski) [reply crates/benda/src/types/mod.rs]: I was thinking about some other methods that could help a type implementation in Benda, I will change the return type to a Result for now. 
The BendType is useful for Generics, I'm using it in:
https://github.com/fglab-tech/benda/blob/28e0bdd1c6b0197441edee2284dc37b0c91589e1/crates/benda/src/types/mod.rs#L18-L27

File: crates/benda/src/types/mod.rs (hunk 4 above)
Comment (vkobinski) [reply crates/benda/src/types/mod.rs:66]: I'll find a way to reference these types in a better way. I don't like this kind of String usage either.

File: crates/benda/src/parser/mod.rs (hunk 6 above)
Comment (vkobinski) [reply crates/benda/src/parser/mod.rs:46]: I tried to remove the excessive use of pattern matching in places in the code where I knew the expected type. I'll find a better way to represent this.

File: crates/benda/src/types/u24.rs (hunk 5 above)
Comment (vkobinski) [reply crates/benda/src/types/u24.rs:82]: The u24, f24 and i24 are proof of concept for now. I'll add the TODOs.

File: crates/benda/src/types/mod.rs (hunk 4 above)
Comment (vkobinski) [reply crates/benda/src/types/mod.rs:66]: The types names are not mapped anywhere in pyo3. I will find a better solution when our subset are better defined. 
I use this From<String> here:
https://github.com/fglab-tech/benda/blob/28e0bdd1c6b0197441edee2284dc37b0c91589e1/crates/benda/src/types/mod.rs#L48

**Pull request merged**